# Limpieza y unión de fuentes

Juntamos las 13 tablas ya limpias (`DF_*_LIMPIO.csv` de buscametas, championchip, carreirasgalegas, ccnorte, cronofinisher, cronorunner, mychip, sportmaniacs, cursescat, iter5, cruzandolameta, youevent y timingsys) en una sola tabla, deduplicamos, derivamos las variables que hacen falta para el análisis y el modelo (año, mes, estación, `total_finishers`, `es_finde`, cruce con la población del INE, índice de carrera popular) y guardamos el resultado en dos ficheros:

- `DF_TODAS.csv`: solo las 12 columnas comunes a las 13 fuentes (`fuente`, `nombre_carrera`, `fecha`, `dia_semana`, `distancia`, `tipo_modalidad`, `publico`, `finisher_d`, `finisher_h`, `municipio`, `comarca`, `provincia`), para quien quiera partir de cero y derivar sus propias columnas — algunos notebooks (p.ej. `Modelo_finishers.ipynb`) ya parten de este fichero.
- `DF_TODAS_UNION.csv`: la misma tabla con todo lo derivado ya calculado, pensada para `Analisis_union.ipynb` y para cualquier notebook de modelado que quiera partir de datos ya listos en vez de recalcularlos.

`raceresult` se queda fuera todavía: su notebook de limpieza depende de un scraping de clasificaciones que solo cubre una pequeña parte de sus 759 eventos (el resto está pendiente de completar en local) y la última ejecución falla al construir `finisher_d`/`finisher_h`, así que se añadirá cuando esa fuente esté lista.

> **Nota (21/9):** se detectó que `finisher_desconocido` (participantes cuyo género no reporta la fuente) se descartaba al quedarnos con las 12 columnas comunes, y que la comprobación de "fila sin resultados" (celda 7) no lo tenía en cuenta. Resultado: para 294 carreras de sportmaniacs (26.436 participantes) y varias más en otras fuentes, TODOS los finishers estaban en esa columna, así que `total_finishers` acababa en 0/NaN y la carrera entera desaparecía del dashboard y del modelo (p. ej. 'Carrera Popular Vega de Triana 2017' o 'III Media Maratón Villa Ducal de Osuna'). Se ha añadido `finisher_desconocido` como 13ª columna (al final, nunca intercalada — ver `Limpieza_outliers.ipynb`), se ha corregido la celda de "sin resultados" y `total_finishers` ahora la suma. `DF_TODAS.csv` / `DF_TODAS_UNION.csv` ya están regenerados con el fix; vuelve a ejecutar este notebook para reproducirlo tú misma.

In [1]:
from pathlib import Path
import pandas as pd
import re

FUENTES = {
    "buscametas": r"../../data/processed/buscametas/DF_BUSCAMETAS_LIMPIO.csv",
    "championchip": r"../../data/processed/championchip/DF_CHAMPIONCHIP_LIMPIO.csv",
    "carreirasgalegas": r"../../data/processed/carreirasgalegas/DF_CARREIRASGALEGAS_LIMPIO.csv",
    "ccnorte": r"../../data/processed/ccnorte/DF_CCNORTE_LIMPIO.csv",
    "cronofinisher": r"../../data/processed/cronofinisher/DF_CRONOFINISHER_LIMPIO.csv",
    "cronorunner": r"../../data/processed/cronorunner/DF_CRONORUNNER_LIMPIO.csv",
    "mychip": r"../../data/processed/mychip/DF_MYCHIP_LIMPIO.csv",
    "sportmaniacs": r"../../data/processed/sportmaniacs/DF_SPORTMANIACS_LIMPIO.csv",
    "cursescat": r"../../data/processed/cursescat/DF_CURSESCAT_LIMPIO.csv",
    "iter5": r"../../data/processed/iter5/DF_ITER5_LIMPIO.csv",
    "cruzandolameta": r"../../data/processed/cruzandolameta/DF_CRUZANDOLAMETA_LIMPIO.csv",
    "youevent": r"../../data/processed/youevent/DF_YOUEVENT_LIMPIO.csv",
    "timingsys": r"../../data/processed/timingsys/DF_TIMINGSYS_LIMPIO.csv",
}

COMUNES = [
    "fuente", "nombre_carrera", "fecha", "dia_semana", "distancia", "tipo_modalidad",
    "publico", "finisher_d", "finisher_h", "municipio", "comarca", "provincia",
]

# FIX (21/9): varias fuentes reportan un tercer grupo de finishers cuyo
# genero no se especifico ("finisher_desconocido"; "finisher_desconegut"
# en buscametas, herencia de un typo catalan en su propia limpieza).
# Hasta ahora esa columna se descartaba aqui al quedarnos solo con las 12
# columnas comunes -- para 294 carreras de sportmaniacs (26.436
# participantes) y varias mas en ccnorte/mychip/cronofinisher/
# cruzandolameta, TODOS los finishers estaban en esa columna, así que la
# carrera entera desaparecia (total_finishers acababa en 0 o NaN) de
# cualquier analisis, del dashboard y del modelo. La capturamos como una
# 13a columna (0 si la fuente no la tiene) en vez de perderla.
_ALIAS_DESCONOCIDO = ["finisher_desconocido", "finisher_desconegut"]

tablas = []
for nombre, ruta in FUENTES.items():
    df_full = pd.read_csv(ruta, encoding="utf-8-sig")
    df = df_full[COMUNES].copy()
    assert (df["fuente"] == nombre).all(), f"'{nombre}': hay filas con 'fuente' distinto, revisar columnas"

    col_desc = next((c for c in _ALIAS_DESCONOCIDO if c in df_full.columns), None)
    if col_desc is not None:
        df["finisher_desconocido"] = df_full[col_desc].fillna(0).astype(float)
    else:
        df["finisher_desconocido"] = 0.0

    print(f"{nombre}: {len(df)} filas" + (f" (finisher_desconocido de '{col_desc}')" if col_desc else " (sin finisher_desconocido)"))
    tablas.append(df)

todas = pd.concat(tablas, ignore_index=True)
todas["fecha"] = pd.to_datetime(todas["fecha"])

print()
print("Total filas:", len(todas))
todas.head()


buscametas: 329 filas (finisher_desconocido de 'finisher_desconegut')
championchip: 2631 filas (sin finisher_desconocido)
carreirasgalegas: 4893 filas (sin finisher_desconocido)
ccnorte: 6220 filas (finisher_desconocido de 'finisher_desconocido')
cronofinisher: 1313 filas (finisher_desconocido de 'finisher_desconocido')
cronorunner: 3539 filas (sin finisher_desconocido)
mychip: 4751 filas (finisher_desconocido de 'finisher_desconocido')
sportmaniacs: 20844 filas (finisher_desconocido de 'finisher_desconocido')
cursescat: 487 filas (sin finisher_desconocido)
iter5: 1834 filas (sin finisher_desconocido)
cruzandolameta: 1718 filas (finisher_desconocido de 'finisher_desconocido')
youevent: 1860 filas (sin finisher_desconocido)
timingsys: 394 filas (sin finisher_desconocido)

Total filas: 50813


,fuente,nombre_carrera,fecha,dia_semana,distancia,tipo_modalidad,publico,finisher_d,finisher_h,municipio,comarca,provincia,finisher_desconocido
0,buscametas,Areta Trail Lasterketa,2026-07-25,Sábado,0.0,trail running,Absoluta/General,17.0,147.0,Areta,NaN,Álava,0.0
1,buscametas,Epic Gravel Ezcaray,2026-07-25,Sábado,0.0,Ciclismo y btt,Absoluta/General,0.0,30.0,Ezcaray,NaN,La Rioja,0.0
2,buscametas,Epic Gravel Ezcaray,2026-07-25,Sábado,93.0,Ciclismo y btt,Absoluta/General,2.0,38.0,Ezcaray,NaN,La Rioja,0.0
3,buscametas,Epic Gravel Ezcaray,2026-07-25,Sábado,57.0,Ciclismo y btt,Absoluta/General,2.0,31.0,Ezcaray,NaN,La Rioja,0.0
4,buscametas,Epic Gravel Ezcaray,2026-07-25,Sábado,42.0,Ciclismo y btt,Absoluta/General,2.0,10.0,Ezcaray,NaN,La Rioja,0.0


In [2]:
# Dos inconsistencias de formato detectadas en EDA_finishers_carreras.ipynb
# (sección 0.4), sin relación con la geografía:
# - 'tipo_modalidad' mezclaba categorías en Mayúscula inicial
#   ("Ciclismo y btt", "Multidisciplina", "Otros") con otras en minúsculas
#   ("marcha", "road running", "trail running") -- son 6 categorías
#   distintas y válidas, pero con un formato inconsistente entre ellas.
# - 'nombre_carrera' traía ~1.260 filas con espacios dobles internos o
#   espacios sobrantes al principio/final, herencia directa del scraping.
#   Alguna fila (p.ej. una de youevent) traía además un salto de línea
#   incrustado en mitad del texto: un '\n' suelto no lo pillaba el
#   `\s{2,}` de antes (hace falta DOS espacios seguidos para disparar),
#   así que sobrevivía sin normalizar -- y al hacer print/display de esa
#   fila, el texto de después del salto aparecía en la columna 0, dando
#   la falsa impresión de ser un valor de 'fuente'. Ahora se normaliza
#   CUALQUIER carácter de espacio en blanco (espacio, tab, salto de
#   línea), no solo los dobles.
antes_tipos = sorted(todas["tipo_modalidad"].dropna().unique())
todas["tipo_modalidad"] = todas["tipo_modalidad"].str.strip().str.capitalize()
print("tipo_modalidad antes:", antes_tipos)
print("tipo_modalidad después:", sorted(todas["tipo_modalidad"].dropna().unique()))

antes_sucio = (
    todas["nombre_carrera"].str.contains(r"\s{2,}|[\t\r\n]", regex=True, na=False)
    | todas["nombre_carrera"].str.match(r"^\s|\s$", na=False)
).sum()
todas["nombre_carrera"] = (
    todas["nombre_carrera"].str.replace(r"\s+", " ", regex=True).str.strip()
)
despues_sucio = (
    todas["nombre_carrera"].str.contains(r"\s{2,}|[\t\r\n]", regex=True, na=False)
    | todas["nombre_carrera"].str.match(r"^\s|\s$", na=False)
).sum()
print(f"\nnombre_carrera con espacios dobles/sobrantes/saltos de línea: {antes_sucio} -> {despues_sucio}")

tipo_modalidad antes: ['Ciclismo y btt', 'Multidisciplina', 'Natación', 'Otros', 'marcha', 'road running', 'trail running']
tipo_modalidad después: ['Ciclismo y btt', 'Marcha', 'Multidisciplina', 'Natación', 'Otros', 'Road running', 'Trail running']

nombre_carrera con espacios dobles/sobrantes/saltos de línea: 1317 -> 0


In [3]:
CCAA_CANONICAS = {
    "andalucia": "Andalucía",
    "aragon": "Aragón",
    "asturias": "Asturias",
    "canarias": "Canarias",
    "cantabria": "Cantabria",
    "castilla y leon": "Castilla y León",
    "castilla la mancha": "Castilla-La Mancha",
    "castilla-la mancha": "Castilla-La Mancha",
    "catalunya": "Cataluña",
    "cataluna": "Cataluña",
    "comunitat valenciana": "Comunidad Valenciana",
    "comunidad valenciana": "Comunidad Valenciana",
    "extremadura": "Extremadura",
    "galicia": "Galicia",
    "illes balears": "Illes Balears",
    "islas baleares": "Illes Balears",
    "la rioja": "La Rioja",
    "murcia": "Murcia",
    "navarra": "Navarra",
    "pais vasco": "País Vasco",
    "país vasco": "País Vasco",
    "comunidad de madrid": "Comunidad de Madrid",
    "madrid": "Comunidad de Madrid",
    "ceuta": "Ceuta",
    "melilla": "Melilla",
    "euskadi": "País Vasco",
    "castile la mancha": "Castilla-La Mancha",
    "castille la mancha": "Castilla-La Mancha",
    "region de murcia": "Murcia",
}

PROVINCIA_CANONICAS = {
    "a coruna": "A Coruña",
    "la coruna": "A Coruña",
    "coruna": "A Coruña",
    "alava": "Álava",
    "araba": "Álava",
    "albacete": "Albacete",
    "alicante": "Alicante",
    "alacant": "Alicante",
    "almeria": "Almería",
    "avila": "Ávila",
    "badajoz": "Badajoz",
    "baleares": "Illes Balears",
    "barcelona": "Barcelona",
    "burgos": "Burgos",
    "caceres": "Cáceres",
    "cadiz": "Cádiz",
    "cantabria": "Cantabria",
    "castellon": "Castellón",
    "castello": "Castellón",
    "ciudad real": "Ciudad Real",
    "cordoba": "Córdoba",
    "cuenca": "Cuenca",
    "girona": "Girona",
    "gerona": "Girona",
    "granada": "Granada",
    "guadalajara": "Guadalajara",
    "guipuzcoa": "Gipuzkoa",
    "gipuzkoa": "Gipuzkoa",
    "huelva": "Huelva",
    "huesca": "Huesca",
    "jaen": "Jaén",
    "leon": "León",
    "lleida": "Lleida",
    "lerida": "Lleida",
    "la rioja": "La Rioja",
    "lugo": "Lugo",
    "madrid": "Madrid",
    "malaga": "Málaga",
    "murcia": "Murcia",
    "navarra": "Navarra",
    "ourense": "Ourense",
    "orense": "Ourense",
    "asturias": "Asturias",
    "asturies": "Asturias",
    "palencia": "Palencia",
    "las palmas": "Las Palmas",
    "pontevedra": "Pontevedra",
    "salamanca": "Salamanca",
    "santa cruz de tenerife": "Santa Cruz de Tenerife",
    "segovia": "Segovia",
    "sevilla": "Sevilla",
    "soria": "Soria",
    "tarragona": "Tarragona",
    "teruel": "Teruel",
    "toledo": "Toledo",
    "valencia": "Valencia",
    "valles": "Valencia",
    "valladolid": "Valladolid",
    "vizcaya": "Vizcaya",
    "bizkaia": "Vizcaya",
    "zamora": "Zamora",
    "zaragoza": "Zaragoza",
    "ceuta": "Ceuta",
    "melilla": "Melilla",
    "baleares (illes)": "Illes Balears",
}

def _normalizar_geo_key(valor):
    if pd.isna(valor):
        return None
    s = str(valor).strip()
    if not s:
        return None
    s = s.replace("-", " ")
    s = re.sub(r"\s+", " ", s)
    s = s.lower()
    return s


def _canonizar_valor(valor, mapping):
    if pd.isna(valor):
        return pd.NA
    s = str(valor).strip()
    if not s:
        return pd.NA
    key = _normalizar_geo_key(s)
    return mapping.get(key, s)



import unicodedata

_VALORES_NO_PROVINCIA = {
    "cualquier lugar del mundo",
    "espana",
    "españa",
}

def _normalizar_geo_key(valor):
    if pd.isna(valor):
        return None
    s = str(valor).strip()
    if not s:
        return None
    s = unicodedata.normalize("NFKD", s)
    s = "".join(ch for ch in s if not unicodedata.combining(ch))
    s = s.replace("-", " ")
    s = re.sub(r"\s+", " ", s)
    s = s.lower()
    return s


def _canonizar_valor(valor, mapping):
    if pd.isna(valor):
        return pd.NA
    s = str(valor).strip()
    if not s:
        return pd.NA
    key = _normalizar_geo_key(s)
    if key in mapping:
        return mapping[key]
    if "/" in key:
        for parte in key.split("/"):
            parte = parte.strip()
            if parte in mapping:
                return mapping[parte]
    return s


def normalitzar_ubicacions(df):
    out = df.copy()

    for col in ["municipio", "comarca", "provincia", "comunidad_autonoma"]:
        if col in out.columns:
            out[col] = out[col].apply(lambda x: None if pd.isna(x) else str(x).strip() or None)

    if "provincia" not in out.columns:
        out["provincia"] = pd.NA
    if "comunidad_autonoma" not in out.columns:
        out["comunidad_autonoma"] = pd.NA

    out["provincia"] = out["provincia"].apply(lambda x: _canonizar_valor(x, PROVINCIA_CANONICAS) if x is not None else pd.NA)
    out["comunidad_autonoma"] = out["comunidad_autonoma"].apply(lambda x: _canonizar_valor(x, CCAA_CANONICAS) if x is not None else pd.NA)

    provincia_key = out["provincia"].fillna("").astype(str).str.strip().str.lower()
    provincia_key = provincia_key.apply(_normalizar_geo_key)

    # FIX (21/9): 9 comunidades autonomas son uniprovinciales -- su unica
    # provincia se llama igual que la propia comunidad (Asturias, Cantabria,
    # Comunidad de Madrid, Murcia, Navarra, La Rioja, Ceuta, Melilla, Illes
    # Balears). Antes, esta mascara trataba CUALQUIER coincidencia con un
    # nombre de CCAA como "una CCAA colada por error en el campo provincia"
    # (el caso real que queremos pillar: una fuente que escribe "Catalunya" o
    # "Comunidad Valenciana" en 'provincia', que si es ambiguo porque esas
    # CCAA tienen varias provincias) y vaciaba 'provincia' tambien para estas
    # 9, aunque el dato fuera correcto -- 3.139 filas se quedaban sin
    # provincia pese a tener municipio conocido (detectado al ver "Ezcaray",
    # La Rioja, sin provincia mientras "Areta", Alava, si la tenia).
    # Excluimos de la mascara todas las claves de CCAA_CANONICAS cuyo valor
    # canonico es una de estas 9 -- así cubre también sinonimos como "Región
    # de Murcia" (43 filas mas que un excluir solo "Murcia" no pillaba).
    _CCAA_UNIPROVINCIALES = {
        "Asturias", "Cantabria", "Comunidad de Madrid", "Murcia", "Navarra",
        "La Rioja", "Ceuta", "Melilla", "Illes Balears",
    }
    _CCAA_UNIPROVINCIAL_KEYS = {
        k for k, v in CCAA_CANONICAS.items() if v in _CCAA_UNIPROVINCIALES
    }
    mask_ccaa_en_provincia = provincia_key.isin(set(CCAA_CANONICAS.keys()) - _CCAA_UNIPROVINCIAL_KEYS)
    if mask_ccaa_en_provincia.any():
        out.loc[mask_ccaa_en_provincia, "comunidad_autonoma"] = out.loc[mask_ccaa_en_provincia, "provincia"]
        out.loc[mask_ccaa_en_provincia, "provincia"] = pd.NA

    mask_no_provincia = provincia_key.isin(_VALORES_NO_PROVINCIA)
    if mask_no_provincia.any():
        out.loc[mask_no_provincia, "provincia"] = pd.NA

    out["provincia"] = out["provincia"].apply(lambda x: _canonizar_valor(x, PROVINCIA_CANONICAS) if x is not None else pd.NA)
    out["comunidad_autonoma"] = out["comunidad_autonoma"].apply(lambda x: _canonizar_valor(x, CCAA_CANONICAS) if x is not None else pd.NA)

    out["comunidad_autonoma"] = out["comunidad_autonoma"].replace("", pd.NA)
    out["provincia"] = out["provincia"].replace("", pd.NA)
    return out

print("Normalitzador de comunitats/províncies aplicat a la unió")

todas = normalitzar_ubicacions(todas)
print("Províncies no nules:", todas["provincia"].notna().sum())
print("Comunitats detectades:", todas["comunidad_autonoma"].notna().sum())


Normalitzador de comunitats/províncies aplicat a la unió
Províncies no nules: 38639
Comunitats detectades: 8163


**Nota (2a regresion detectada y corregida, 21/9):** `normalitzar_ubicacions()` vaciaba `provincia` para las 9 comunidades autonomas uniprovinciales (su provincia se llama igual que la CCAA, o un sinonimo como "Región de Murcia") cada vez que una fuente escribia correctamente ese nombre -- confundiendolo con el caso real que la mascara queria detectar (una CCAA multiprovincial, tipo "Catalunya", colada por error en el campo `provincia`). 3.139 filas de Asturias, Illes Balears, Cantabria, Murcia, Navarra, La Rioja, Ceuta y Melilla se quedaban con `provincia` vacia pese a tener `municipio` conocido -- detectado al revisar por que "Ezcaray" (La Rioja) aparecia sin provincia mientras "Areta" (Alava) si la tenia. Corregido excluyendo, de la mascara, todas las claves de `CCAA_CANONICAS` cuyo valor es una de estas 9 comunidades uniprovinciales (cubre tambien sinonimos como "Región de Murcia"); `provincia_ine`/`total_poblacion` no empeoran -- de hecho suben un poco (84.9% -> 91.6%), porque ya no dependen del fallback de `CCAA_UNIPROVINCIAL` en la celda del cruce con el INE para resolverse.

In [4]:
# Unificación de grafías de 'municipio' (EDA, sección 0.4): mismo
# municipio escrito con mayúsculas o acentos distintos según la fila
# ("LLEIDA" / "Lleida", "AITONA" / "Aitona"...). Nos quedamos, para cada
# municipio, con la grafía más frecuente en los propios datos.
def _clave_municipio(valor):
    if pd.isna(valor):
        return None
    return unicodedata.normalize("NFKD", str(valor)).encode("ascii", "ignore").decode("ascii").upper().strip()

_clave = todas["municipio"].apply(_clave_municipio)
_grafia_mas_frecuente = (
    todas.dropna(subset=["municipio"])
    .groupby(_clave.loc[todas["municipio"].notna()])["municipio"]
    .agg(lambda s: s.value_counts().idxmax())
)
_municipios_con_variantes = (
    todas.dropna(subset=["municipio"]).groupby(_clave.loc[todas["municipio"].notna()])["municipio"]
    .nunique()
)
print(f"Municipios con más de una grafía antes de unificar: {(_municipios_con_variantes > 1).sum()}")

todas["municipio"] = _clave.map(_grafia_mas_frecuente)

_variantes_despues = (
    todas.dropna(subset=["municipio"]).groupby(_clave.loc[todas["municipio"].notna()])["municipio"].nunique()
)
print(f"Municipios con más de una grafía después de unificar: {(_variantes_despues > 1).sum()}")

Municipios con más de una grafía antes de unificar: 124
Municipios con más de una grafía después de unificar: 0


In [5]:
# Un mismo evento puede haber sido capturado por más de una fuente (p.ej.
# una carrera popular que sale tanto en sportmaniacs como en cronofinisher)
# — sin quitar esto, esas carreras se cuentan dos veces. Detectamos
# duplicados por nombre normalizado (sin acentos/mayúsculas/puntuación,
# para no fallar por "1ª" vs "I" o tildes) + fecha exacta. Cuando el mismo
# evento aparece en varias fuentes, nos quedamos solo con la fuente que
# tiene más finishers sumados para ese evento — normalmente porque la otra
# fuente solo registró el evento sin resultados (filas en 0) o con menos
# categorías capturadas.
import unicodedata


def _normalizar_nombre(texto):
    if pd.isna(texto):
        return None
    t = unicodedata.normalize("NFKD", str(texto)).encode("ascii", "ignore").decode("ascii").upper()
    return re.sub(r"[^A-Z0-9]+", " ", t).strip()


todas["_nombre_norm"] = todas["nombre_carrera"].apply(_normalizar_nombre)
todas["_total_tmp"] = todas["finisher_d"] + todas["finisher_h"]

resumen = todas.groupby(["_nombre_norm", "fecha", "fuente"])["_total_tmp"].sum().reset_index()
n_fuentes = resumen.groupby(["_nombre_norm", "fecha"])["fuente"].transform("nunique")
resumen_dup = resumen[n_fuentes > 1]

idx_ganador = resumen_dup.groupby(["_nombre_norm", "fecha"])["_total_tmp"].idxmax()
ganador = resumen_dup.loc[idx_ganador, ["_nombre_norm", "fecha", "fuente"]].rename(
    columns={"fuente": "_fuente_ganadora"}
)

antes = len(todas)
todas = todas.merge(ganador, on=["_nombre_norm", "fecha"], how="left")
es_evento_duplicado = todas["_fuente_ganadora"].notna()
filas_a_quitar = es_evento_duplicado & (todas["fuente"] != todas["_fuente_ganadora"])

print(f"Eventos duplicados entre fuentes: {ganador.shape[0]}")
print(f"Filas eliminadas (de la fuente no ganadora): {filas_a_quitar.sum()}")
print(f"  - con 0 finishers (fuente que no llegó a capturar resultados): "
      f"{(filas_a_quitar & (todas['_total_tmp'] == 0)).sum()}")
print(f"  - con datos, pero menos completos que la fuente ganadora: "
      f"{(filas_a_quitar & (todas['_total_tmp'] > 0)).sum()}")
print()
print("Filas eliminadas, por fuente:")
print(todas.loc[filas_a_quitar, "fuente"].value_counts())

todas = todas.loc[~filas_a_quitar].drop(columns=["_nombre_norm", "_total_tmp", "_fuente_ganadora"])
print()
print(f"Total filas: {antes} -> {len(todas)}")

Eventos duplicados entre fuentes: 31
Filas eliminadas (de la fuente no ganadora): 68
  - con 0 finishers (fuente que no llegó a capturar resultados): 39
  - con datos, pero menos completos que la fuente ganadora: 28

Filas eliminadas, por fuente:
fuente
cronofinisher       25
mychip              25
cruzandolameta      10
sportmaniacs         3
carreirasgalegas     2
ccnorte              2
youevent             1
Name: count, dtype: int64

Total filas: 50813 -> 50745


In [6]:
# Además de eventos repetidos entre fuentes, hay filas totalmente
# idénticas dentro de una misma fuente (mismo evento, categoría y hasta el
# mismo número de finishers) — repartidas entre 7 fuentes distintas sin
# que ninguna concentre la mayoría, así que parece un artefacto general de
# scraping (páginas o categorías leídas dos veces) y no un fallo de una
# fuente en concreto. Al ser copias exactas, quedarnos con una sola no
# pierde ninguna información real.
#
# FIX (21/9): la clave incluye ahora tambien finisher_desconocido -- antes
# dos filas que coincidian en finisher_d/finisher_h pero tenian un
# finisher_desconocido distinto se trataban como "copia exacta" y se
# perdia esa informacion real al quedarnos con una sola.
cols_evento = [
    "fuente", "nombre_carrera", "fecha", "distancia", "tipo_modalidad", "publico",
    "finisher_d", "finisher_h", "finisher_desconocido",
]
antes = len(todas)
print("Filas idénticas dentro de la misma fuente, por fuente:")
print(todas.loc[todas.duplicated(subset=cols_evento, keep=False), "fuente"].value_counts())

todas = todas.drop_duplicates(subset=cols_evento).reset_index(drop=True)
print()
print(f"Total filas: {antes} -> {len(todas)}")


Filas idénticas dentro de la misma fuente, por fuente:
fuente
sportmaniacs      468
ccnorte           396
mychip            338
championchip      100
cronofinisher      38
buscametas         32
cruzandolameta     10
Name: count, dtype: int64

Total filas: 50745 -> 49953


In [7]:
# cronofinisher (79% de sus filas) y carreirasgalegas (66%) tienen una
# proporción de filas en 0/0 mucho más alta que el resto de fuentes, y no
# se explica por categorías especiales (pasa igual en Absoluta/General) ni
# solo por carreras que aún no se han disputado (hay filas en 0 desde
# 2015/2019) — son categorías que existieron pero de las que nunca se
# llegó a capturar el resultado real. Las marcamos como NaN, no 0, igual
# que ya se hace con el centinela de "distancia" — así no se cuentan como
# "carrera real de 0 personas" ni en las estadísticas ni en un futuro modelo.
#
# Actualización (EDA_finishers_carreras.ipynb, sección 0.2): restringir
# esto a solo estas 2 fuentes dejaba fuera casi 3.000 filas con
# exactamente el mismo patrón en otras 6 fuentes (ccnorte, mychip,
# youevent, sportmaniacs, buscametas, championchip), todas de carreras ya
# disputadas y repartidas desde 2007 -- no un caso puntual de estas 2
# fuentes. Generalizamos el criterio a todas las fuentes: fila en 0/0 de
# una carrera cuya fecha ya ha pasado.
#
# FIX (21/9): añadimos "& finisher_desconocido == 0" -- sin esto, una
# carrera real cuyos participantes SOLO estaban en finisher_desconocido
# (género no especificado por la fuente) se marcaba igualmente como "sin
# resultados" y se perdía entera (ver nota de la celda 1).
_sin_resultados = (
    (todas["finisher_d"] == 0)
    & (todas["finisher_h"] == 0)
    & (todas["finisher_desconocido"] == 0)
    & (todas["fecha"] <= pd.Timestamp.today().normalize())
)

print(f"Filas marcadas como sin resultados (antes 0/0/0, carrera ya disputada): {_sin_resultados.sum()}")
print(todas.loc[_sin_resultados, "fuente"].value_counts())

todas["finisher_d"] = todas["finisher_d"].astype("float64")
todas["finisher_h"] = todas["finisher_h"].astype("float64")
todas.loc[_sin_resultados, ["finisher_d", "finisher_h"]] = float("nan")


Filas marcadas como sin resultados (antes 0/0/0, carrera ya disputada): 6732
fuente
carreirasgalegas    3933
ccnorte             1205
mychip              1052
youevent             501
championchip          23
buscametas            16
sportmaniacs           1
timingsys              1
Name: count, dtype: int64


In [8]:
# La fusión "Cadete/Juvenil" -> "Infantil" ya no se hace aquí: se movió a
# cada Limpieza_<fuente>.ipynb (carreirasgalegas y ccnorte ya lo hacían así
# en origen; el resto se cambió para que 'Infantil' signifique lo mismo en
# las 13 fuentes desde su propia limpieza, no como parche al unir). Solo
# comprobamos aquí que ninguna fuente se haya quedado atrás.
assert "Cadete/Juvenil" not in todas["publico"].unique(), (
    "alguna fuente todavia distingue Cadete/Juvenil de Infantil - revisar su Limpieza_<fuente>.ipynb"
)
print(todas["publico"].value_counts())

publico
Absoluta/General     35327
Infantil             12611
Otros                  623
Mayores/Veteranos      542
Equipos                510
Elite                  340
Name: count, dtype: int64


In [9]:
# distancia / municipio / comarca / provincia pueden venir vacíos según la
# fuente; finisher_d/finisher_h también, pero solo en cronofinisher y
# carreirasgalegas (las filas sin resultados reales que acabamos de marcar
# como NaN) -- el resto de columnas no debería tener ningún nulo.
print("Filas por fuente:")
print(todas["fuente"].value_counts())
print()

print("Nulos por columna:")
print(todas.isna().sum())
print()

print("Rango de fechas:", todas["fecha"].min(), "->", todas["fecha"].max())

Filas por fuente:
fuente
sportmaniacs        20563
ccnorte              5984
carreirasgalegas     4891
mychip               4539
cronorunner          3539
championchip         2579
youevent             1859
iter5                1834
cruzandolameta       1703
cronofinisher        1268
cursescat             487
timingsys             394
buscametas            313
Name: count, dtype: int64

Nulos por columna:
fuente                      0
nombre_carrera              0
fecha                       0
dia_semana                  0
distancia                   0
tipo_modalidad              0
publico                     0
finisher_d               8576
finisher_h               8576
municipio                3439
comarca                 23807
provincia               11945
finisher_desconocido        0
comunidad_autonoma      41937
dtype: int64

Rango de fechas: 2003-11-16 00:00:00 -> 2026-09-13 00:00:00


In [10]:
SALIDA = Path("../../data/processed/union/DF_TODAS.csv")
todas.to_csv(SALIDA, index=False, encoding="utf-8-sig")
print("Guardado en", SALIDA)

Guardado en ../../data/processed/union/DF_TODAS.csv


## Variables derivadas y cruce con población

Antes de meternos en el análisis añadimos las columnas que vamos a necesitar todo el rato: año, mes y estación a partir de la fecha, el total de finishers por fila, si la carrera es de fin de semana, y la población del municipio del año de la carrera y su superficie/densidad (cruzando con la serie histórica del INE de `Limpieza_INE.ipynb`). Así las tenemos ya listas como variables numéricas para el univariable, la correlación y el cruce con los objetivos.

In [11]:
MES_A_ESTACION = {
    12: "Invierno", 1: "Invierno", 2: "Invierno",
    3: "Primavera", 4: "Primavera", 5: "Primavera",
    6: "Verano", 7: "Verano", 8: "Verano",
    9: "Otoño", 10: "Otoño", 11: "Otoño",
}

todas["anyo"] = todas["fecha"].dt.year
todas["mes"] = todas["fecha"].dt.month
todas["estacion"] = todas["mes"].map(MES_A_ESTACION)
# FIX (21/9): incluye finisher_desconocido -- ver nota de la celda 1.
todas["total_finishers"] = todas["finisher_d"] + todas["finisher_h"] + todas["finisher_desconocido"].fillna(0)
todas["es_finde"] = todas["dia_semana"].isin(["Sábado", "Domingo"])

todas[["fecha", "anyo", "mes", "estacion", "total_finishers", "es_finde"]].head()


,fecha,anyo,mes,estacion,total_finishers,es_finde
0,2026-07-25,2026,7,Verano,164.0,True
1,2026-07-25,2026,7,Verano,30.0,True
2,2026-07-25,2026,7,Verano,40.0,True
3,2026-07-25,2026,7,Verano,33.0,True
4,2026-07-25,2026,7,Verano,12.0,True


### Estimar el género para las filas con total conocido pero desglose desconocido (29/9)

Hay filas donde `finisher_d=finisher_h=0` pero `finisher_desconocido>0` -- no son una
carrera real con cero participantes (esas ya se marcaron como `NaN` más arriba, en la
celda de "sin resultados"): son carreras que **sí** tuvieron participantes, solo que la
fuente no desglosó el género. `total_finishers` ya las cuenta bien desde el fix del
21/9 (`total_finishers = finisher_d + finisher_h + finisher_desconocido`), pero
`finisher_d`/`finisher_h` por separado siguen sin decir nada útil de esas filas (1.265
filas, un 3,2% del dataset).

En vez de dejarlas así sin más, se estima cómo se habría repartido ese total por
género: se calcula el % de mujeres observado en carreras **similares y del mismo año**
(`tipo_modalidad` + `publico` + `anyo`, usando solo filas de género realmente conocido)
y se aplica ese % al `total_finishers` de la fila. Si el grupo específico no tiene
suficientes participantes con género conocido como para fiarse del porcentaje (umbral:
100), se retrocede a un grupo más amplio (se quita primero `anyo`, luego `publico`, y en
último caso solo se exige `tipo_modalidad`) -- mismo criterio de retroceso por niveles
que ya usa `Modelo_finishers.ipynb` para el historial por comparables.

**Importante:**
- Esto **nunca sobreescribe** `finisher_d`/`finisher_h` reales (que en este notebook
  siguen siendo `0`/`0` para estas filas -- el fix que las pasa a `NaN` vive en
  `Limpieza_outliers.ipynb`, a propósito, porque tiene que aplicarse después de la
  winsorización de ese notebook). El resultado se guarda en columnas nuevas,
  `finisher_d_est`/`finisher_h_est`, para que quede siempre trazable qué es un dato
  observado y qué es una estimación. `nivel_pct_genero` guarda qué tan específico fue
  el grupo usado (0 = modalidad+público+año, el más fiable; 3 = solo el % global de
  todo el dataset, el menos fiable).
- **Uso previsto para el EDA/dashboard:** al calcular el % de mujeres agregado por
  año/modalidad, usar `finisher_d.fillna(finisher_d_est)` (e igual para `finisher_h`)
  recupera estas 1.265 filas en vez de perderlas silenciosamente -- pero conviene dejar
  siempre una nota metodológica de que una parte del desglose es estimada, no observada
  (ver `claude/fix_imputacion_genero_desconocido.md` para el detalle completo).
- **Uso probado para el modelo de ML:** se comprobó (ver `Modelo_finishers.ipynb`,
  sección "Modelo v2") si añadir estas filas imputadas al **entrenamiento** (nunca a la
  evaluación, que se queda solo con género real) mejora las métricas. Dentro del universo
  del modelo (running/trail + público general) solo una fracción pequeña de estas filas
  es recuperable y cae en años de train, y el efecto en MAE/RMSE/R²/error relativo
  resultó nulo -- dentro del ruido de un entrenamiento a otro. Conclusión: el modelo v2
  sigue entrenándose y evaluándose solo con género real; `finisher_d_est`/
  `finisher_h_est` no se usan como target, mejor reservarlas para el EDA y el dashboard.

In [ ]:
import numpy as np

NIVELES_PCT_GENERO = [
    ["tipo_modalidad", "publico", "anyo"],
    ["tipo_modalidad", "publico"],
    ["tipo_modalidad"],
]
MIN_PARTICIPANTES_PCT = 100  # por debajo de esto, no fiarse del % del grupo y retroceder

# Filas "recuperables": el patron del falso 0/0 (ver nota de la celda "sin resultados"
# mas arriba) -- total conocido (via finisher_desconocido), genero no. A diferencia de
# Limpieza_outliers.ipynb, aqui finisher_d/finisher_h todavia son 0/0 literales (el fix
# que los pasa a NaN vive en Outliers, despues de la winsorizacion), asi que se detectan
# por el patron 0/0 + finisher_desconocido>0 en vez de por NaN.
recuperable = (
    (todas["finisher_d"] == 0) & (todas["finisher_h"] == 0) & (todas["finisher_desconocido"] > 0)
)
genero_conocido = ~recuperable & todas["finisher_d"].notna() & todas["finisher_h"].notna()

print(f"Filas con genero conocido: {genero_conocido.sum()} de {len(todas)} ({100 * genero_conocido.mean():.1f}%)")
print(f"Filas recuperables por estimacion (total conocido, genero no): {recuperable.sum()}")

pct_estim = pd.Series(np.nan, index=todas.index)
nivel_pct = pd.Series(len(NIVELES_PCT_GENERO), index=todas.index)
pendiente = recuperable.copy()

for nivel, claves in enumerate(NIVELES_PCT_GENERO):
    agregados = todas.loc[genero_conocido].groupby(claves)[["finisher_d", "finisher_h"]].sum()
    agregados["n_participantes"] = agregados["finisher_d"] + agregados["finisher_h"]
    agregados["pct_mujeres"] = agregados["finisher_d"] / agregados["n_participantes"]
    validos = agregados.loc[agregados["n_participantes"] >= MIN_PARTICIPANTES_PCT, "pct_mujeres"]

    filas_pendientes = todas.loc[pendiente, claves]
    match = filas_pendientes.merge(validos.rename("pct_mujeres"), left_on=claves, right_index=True, how="left")
    idx_resueltas = match.index[match["pct_mujeres"].notna()]

    pct_estim.loc[idx_resueltas] = match.loc[idx_resueltas, "pct_mujeres"]
    nivel_pct.loc[idx_resueltas] = nivel
    pendiente.loc[idx_resueltas] = False

# Fallback final: % global de mujeres entre todas las filas de genero conocido.
pct_global = (
    todas.loc[genero_conocido, "finisher_d"].sum()
    / (todas.loc[genero_conocido, "finisher_d"].sum() + todas.loc[genero_conocido, "finisher_h"].sum())
)
pct_estim.loc[pendiente] = pct_global

todas["finisher_d_est"] = np.nan
todas["finisher_h_est"] = np.nan
todas.loc[recuperable, "finisher_d_est"] = todas.loc[recuperable, "total_finishers"] * pct_estim.loc[recuperable]
todas.loc[recuperable, "finisher_h_est"] = todas.loc[recuperable, "total_finishers"] * (1 - pct_estim.loc[recuperable])
todas["nivel_pct_genero"] = np.nan
todas.loc[recuperable, "nivel_pct_genero"] = nivel_pct.loc[recuperable]

print(f"\nfinisher_d_est/finisher_h_est calculados para {recuperable.sum()} filas.")
print("Nivel usado para el % de mujeres (0=modalidad+publico+anyo, 1=modalidad+publico, "
      "2=modalidad, 3=fallback global):")
print(nivel_pct.loc[recuperable].value_counts().sort_index())

In [12]:
import glob
import re
import unicodedata

def _despostponer_articulo(texto_norm):
    """Pasa el artículo pospuesto entre paréntesis del nomenclátor ("CORUÑA (A)",
    "POBLA DE FARNALS (LA)") delante, que es como se escribe en las fuentes ("A
    CORUÑA", "LA POBLA DE FARNALS") -- sin espacio tras el apóstrofo elidido de
    "L'", con espacio para el resto de artículos. Recibe texto ya en mayúsculas
    sin acentos (de `normalizar()` o de `municipio_norm`/`nombre_ine` del INE).
    """
    if texto_norm is None:
        return None
    m = re.match(r"^(.*?)\s*\((L'|LA|EL|LOS|LAS|ELS|LES|A|O|AS|OS)\)$", texto_norm)
    if not m:
        return texto_norm
    nombre, articulo = m.groups()
    sep = "" if articulo.endswith("'") else " "
    return f"{articulo}{sep}{nombre}".strip()


def normalizar(texto):
    """Mayúsculas y sin acentos, para poder cruzar texto libre con el nomenclátor del INE.

    También normaliza dos formatos propios del nomenclátor: los nombres dobles
    oficiales tipo "ALBORAIA/ALBORAYA" (se homogeniza el espaciado alrededor de
    la barra) y el artículo pospuesto entre paréntesis tipo "ALCÚDIA (L')" o
    "CORUÑA (A)" (se pasa a "L'ALCUDIA"/"A CORUNA", que es como suele venir
    escrito en las fuentes — sin espacio tras el apóstrofo elidido de "L'",
    con espacio para el resto de artículos).
    """
    if pd.isna(texto):
        return None
    t = unicodedata.normalize("NFKD", str(texto)).encode("ascii", "ignore").decode("ascii").upper().strip()
    t = re.sub(r"\s*/\s*", "/", t)
    return _despostponer_articulo(t)

# Las comunidades multiprovinciales (Galicia, Catalunya, Andalucía, la
# Comunidad Valenciana, Castilla y León, Castilla-La Mancha, Aragón) no se
# pueden asignar a una sola provincia del INE con el nombre de la comunidad
# solo, así que además de la tabla de provincias guardamos, para cada una de
# estas comunidades, las provincias que la componen: si el municipio tiene un
# nombre único dentro de esas provincias, podemos igualmente encontrar su
# población sin ambigüedad (ver el fallback más abajo).
PROVINCIA_INE = {
    1: ["Álava", "Araba", "Araba/Álava"], 2: ["Albacete"], 3: ["Alicante", "Alacant"],
    4: ["Almería"], 5: ["Ávila"], 6: ["Badajoz"], 7: ["Baleares", "Illes Balears", "Baleares (Illes)"],
    8: ["Barcelona"], 9: ["Burgos"], 10: ["Cáceres"], 11: ["Cádiz"], 12: ["Castellón", "Castelló"],
    13: ["Ciudad Real"], 14: ["Córdoba"], 15: ["La Coruña", "A Coruña"], 16: ["Cuenca"],
    17: ["Girona", "Gerona"], 18: ["Granada"], 19: ["Guadalajara"], 20: ["Guipúzcoa", "Gipuzkoa"],
    21: ["Huelva"], 22: ["Huesca"], 23: ["Jaén"], 24: ["León"], 25: ["Lleida", "Lérida"],
    26: ["La Rioja"], 27: ["Lugo"], 28: ["Madrid", "Comunidad de Madrid"], 29: ["Málaga"],
    30: ["Murcia", "Región de Murcia"], 31: ["Navarra"], 32: ["Ourense", "Orense"],
    33: ["Asturias", "Asturies", "Asturias / Asturies"], 34: ["Palencia"], 35: ["Las Palmas"],
    36: ["Pontevedra"], 37: ["Salamanca"], 38: ["Santa Cruz de Tenerife"], 39: ["Cantabria"],
    40: ["Segovia"], 41: ["Sevilla"], 42: ["Soria"], 43: ["Tarragona"], 44: ["Teruel"],
    45: ["Toledo"], 46: ["Valencia", "València"], 47: ["Valladolid"], 48: ["Vizcaya", "Bizkaia"],
    49: ["Zamora"], 50: ["Zaragoza"], 51: ["Ceuta"], 52: ["Melilla"],
}
PROVINCIA_A_CODIGO = {normalizar(n): c for c, ns in PROVINCIA_INE.items() for n in ns}

CCAA_MULTIPROVINCIA = {
    "Galicia": [15, 27, 32, 36],
    "Catalunya": [8, 17, 25, 43],
    "Cataluña": [8, 17, 25, 43],
    "Andalucía": [4, 11, 14, 18, 21, 23, 29, 41],
    "Comunitat Valenciana": [3, 12, 46],
    "Comunidad Valenciana": [3, 12, 46],
    "País Vasco": [1, 20, 48],
    "Aragón": [22, 44, 50],
    "Castilla-La Mancha": [2, 13, 16, 19, 45],
    "Castile-La Mancha": [2, 13, 16, 19, 45],
    "Castille-La Mancha": [2, 13, 16, 19, 45],
    "Castilla y León": [5, 9, 24, 34, 37, 40, 42, 47, 49],
    "Extremadura": [6, 10],
    "Canarias": [35, 38],
}
CCAA_A_PROVINCIAS = {normalizar(k): v for k, v in CCAA_MULTIPROVINCIA.items()}

# Las comunidades uniprovinciales (su territorio es exactamente una
# provincia) sí se pueden resolver sin ninguna ambigüedad ni comprobación de
# unicidad -- a diferencia de las multiprovinciales de arriba, aquí no hace
# falta mirar el municipio para nada, el nombre de la comunidad ya dice la
# provincia. Son las que se quedaban sin cruzar en cuanto `provincia` traía
# el nombre de la comunidad en vez de la provincia (p.ej. "Cantabria",
# "Región de Murcia") y `normalitzar_ubicacions()` lo pasaba a
# `comunidad_autonoma`, vaciando `provincia`.
CCAA_UNIPROVINCIAL = {
    "Asturias": 33, "Principado de Asturias": 33,
    "Cantabria": 39,
    "Madrid": 28, "Comunidad de Madrid": 28,
    "Murcia": 30, "Región de Murcia": 30,
    "Navarra": 31, "Comunidad Foral de Navarra": 31,
    "La Rioja": 26,
    "Illes Balears": 7, "Islas Baleares": 7, "Baleares": 7,
    "Ceuta": 51,
    "Melilla": 52,
}
CCAA_UNIPROVINCIAL_A_CODIGO = {normalizar(k): v for k, v in CCAA_UNIPROVINCIAL.items()}

# --- Población y superficie: ahora vienen de Limpieza_INE.ipynb -----------
# DF_INE_POBLACION_HISTORICO.csv trae una fila por (municipio, año) en vez de
# un único año fijo como el DF_INE_Poblaciones_España.xlsx anterior, así que
# el cruce pasa a hacerse por (municipio_norm, provincia_ine, año). Para
# carreras de años sin padrón todavía (el año en curso) o anteriores al primer
# año disponible, usamos el año disponible más cercano -- mismo criterio de
# "quedarnos con el último año disponible" que ya veníamos aplicando en otros
# sitios del proyecto. La superficie no depende del año (los límites
# municipales apenas cambian), así que se cruza aparte y sirve para la
# densidad de población.
INE_DIR = Path("../../data/processed/INE")
ine_historico = pd.read_csv(INE_DIR / "DF_INE_POBLACION_HISTORICO.csv")
ine_superficie = pd.read_csv(INE_DIR / "DF_INE_SUPERFICIE_MUNICIPIOS.csv")

# `Limpieza_INE.ipynb` normaliza mayúsculas/acentos pero no despospone el
# artículo entre paréntesis del nomenclátor (esa conversión se añadió después,
# aquí, solo para el lado de `todas`) -- sin este paso ~15.500 filas del INE se
# quedan con nombres tipo "POBLA DE FARNALS (LA)" que nunca cruzan contra
# "LA POBLA DE FARNALS", la forma en que vienen las fuentes. Se aplica antes de
# expandir los alias dobles porque algún nombre combina ambos formatos (p.ej.
# "BENITACHELL/POBLE NOU DE BENITATXELL (EL)").
ine_historico["municipio_norm"] = ine_historico["municipio_norm"].apply(_despostponer_articulo)
ine_superficie["municipio_norm"] = ine_superficie["municipio_norm"].apply(_despostponer_articulo)

ANYO_MIN_DISPONIBLE = ine_historico["anyo"].min()
ANYO_MAX_DISPONIBLE = ine_historico["anyo"].max()

# El INE usa nombres dobles oficiales tipo "ALBORAIA/ALBORAYA"; los añadimos
# también como dos filas independientes (una por cada mitad) para poder
# cruzar aunque una fuente solo use una de las dos formas del nombre. Se
# aplica igual a población (por año) y a superficie (una sola fila).
def expandir_alias_dobles(df):
    dobles = df[df["municipio_norm"].str.contains("/", na=False)]
    if dobles.empty:
        return df
    alias = [
        {**fila.to_dict(), "municipio_norm": mitad.strip()}
        for _, fila in dobles.iterrows()
        for mitad in fila["municipio_norm"].split("/")
    ]
    return pd.concat([df, pd.DataFrame(alias)], ignore_index=True)

ine_historico = expandir_alias_dobles(ine_historico)
ine_historico = ine_historico.drop_duplicates(subset=["municipio_norm", "provincia_ine", "anyo"], keep="first")

ine_superficie = expandir_alias_dobles(ine_superficie)
ine_superficie = ine_superficie.drop_duplicates(subset=["municipio_norm", "provincia_ine"], keep="first")

todas["municipio_norm"] = todas["municipio"].apply(normalizar)
todas["provincia_norm"] = todas["provincia"].apply(normalizar)
todas["comunidad_autonoma_norm"] = todas["comunidad_autonoma"].apply(normalizar)
def resolver_provincia_directa(provincia_norm):
    """Prueba el nombre completo y, si trae barra ("VALENCIA/VALENCIA",
    "ASTURIES" suelto no la tiene pero "ALACANT/ALICANTE" sí), cada mitad
    por separado contra el diccionario de provincias."""
    if provincia_norm is None:
        return None
    for parte in provincia_norm.split("/"):
        if parte in PROVINCIA_A_CODIGO:
            return PROVINCIA_A_CODIGO[parte]
    return None


todas["provincia_ine"] = todas["provincia_norm"].apply(resolver_provincia_directa)

# Antes del fallback multiprovincial (que necesita comprobar unicidad del
# municipio), resolvemos ya sin ambigüedad las filas cuya única pista es una
# comunidad uniprovincial -- no hay nada que comprobar, el nombre de la
# comunidad ya fija la provincia.
_sin_provincia = todas["provincia_ine"].isna()
todas.loc[_sin_provincia, "provincia_ine"] = todas.loc[_sin_provincia, "comunidad_autonoma_norm"].map(
    CCAA_UNIPROVINCIAL_A_CODIGO
)
print(
    "Filas resueltas por comunidad uniprovincial (antes del cruce): "
    f"{int((_sin_provincia & todas['provincia_ine'].notna()).sum())}"
)

# Año de cruce: el de la propia carrera, salvo que caiga fuera del rango que
# cubre el INE -- se acerca al extremo disponible más próximo en vez de dejarlo
# sin cruzar.
todas["anyo_cruce"] = todas["anyo"].clip(lower=ANYO_MIN_DISPONIBLE, upper=ANYO_MAX_DISPONIBLE)
n_ajustados = int((todas["anyo_cruce"] != todas["anyo"]).sum())
print(
    f"Filas cuyo año de carrera queda fuera de {ANYO_MIN_DISPONIBLE}-{ANYO_MAX_DISPONIBLE} "
    f"(se cruza con el año disponible más cercano): {n_ajustados}"
)

todas = todas.merge(
    ine_historico[["municipio_norm", "provincia_ine", "anyo", "total_poblacion", "poblacion_h", "poblacion_d"]],
    left_on=["municipio_norm", "provincia_ine", "anyo_cruce"],
    right_on=["municipio_norm", "provincia_ine", "anyo"],
    how="left",
    suffixes=("", "_ine"),
)
todas = todas.drop(columns=["anyo_ine"], errors="ignore")

todas = todas.merge(
    ine_superficie[["municipio_norm", "provincia_ine", "superficie_km2"]],
    on=["municipio_norm", "provincia_ine"],
    how="left",
)
todas["densidad_poblacion"] = todas["total_poblacion"] / todas["superficie_km2"]

# Fallback para comunidades multiprovinciales: cuando 'provincia' es en
# realidad el nombre de una comunidad autónoma con varias provincias (o el
# cruce anterior no encontró nada), buscamos el municipio dentro de las
# provincias de esa comunidad, para el año de cruce correspondiente. Solo lo
# damos por válido si el nombre del municipio es único en toda la comunidad
# ese año, para no arriesgarnos a cruzar con la provincia equivocada cuando
# el nombre se repite.
#
# El candidato de comunidad puede venir de 'provincia' (fuentes que meten ahí
# directamente el nombre de la comunidad, p.ej. iter5) o de
# 'comunidad_autonoma' -- `normalitzar_ubicacions()` ya vacía 'provincia' y
# traslada el valor a 'comunidad_autonoma' en cuanto detecta que lo que traía
# la fuente era en realidad una comunidad autónoma (p.ej. cronorunner, que
# llega con 'provincia' = "Comunitat Valenciana" en el 85% de sus filas), así
# que si solo miráramos 'provincia_norm' aquí ese caso nunca encontraría
# candidatas y se quedaría sin cruzar casi entero.
_lookup = {}
for (municipio, anyo_fila), grupo in ine_historico.groupby(["municipio_norm", "anyo"]):
    for _, fila in grupo.iterrows():
        _lookup.setdefault(municipio, {}).setdefault(anyo_fila, {})[fila["provincia_ine"]] = fila

_superficie_lookup = {}
for _, fila in ine_superficie.iterrows():
    _superficie_lookup.setdefault(fila["municipio_norm"], {})[fila["provincia_ine"]] = fila["superficie_km2"]

_pendientes = todas.index[todas["total_poblacion"].isna()]
_resueltos_ccaa = 0
def candidatas_ccaa(provincia_norm, ccaa_norm):
    encontradas = set()
    for valor in (provincia_norm, ccaa_norm):
        if valor is None:
            continue
        for parte in valor.split("/"):
            encontradas.update(CCAA_A_PROVINCIAS.get(parte, []))
    return list(encontradas) if encontradas else None


for idx in _pendientes:
    candidatas = candidatas_ccaa(todas.at[idx, "provincia_norm"], todas.at[idx, "comunidad_autonoma_norm"])
    if not candidatas:
        continue
    anyo_fila = todas.at[idx, "anyo_cruce"]
    opciones = _lookup.get(todas.at[idx, "municipio_norm"], {}).get(anyo_fila)
    if not opciones:
        continue
    encontrados = {p: v for p, v in opciones.items() if p in candidatas}
    if len(encontrados) == 1:
        prov_ine, fila = list(encontrados.items())[0]
        todas.loc[idx, ["provincia_ine", "total_poblacion", "poblacion_h", "poblacion_d"]] = [
            prov_ine, fila["total_poblacion"], fila["poblacion_h"], fila["poblacion_d"]
        ]
        superficie_val = _superficie_lookup.get(todas.at[idx, "municipio_norm"], {}).get(prov_ine)
        if superficie_val is not None:
            todas.loc[idx, "superficie_km2"] = superficie_val
            todas.loc[idx, "densidad_poblacion"] = fila["total_poblacion"] / superficie_val
        _resueltos_ccaa += 1

# Último fallback: sin ninguna pista de provincia ni comunidad autónoma, pero
# el nombre del municipio es único en todo el nomenclátor de España (ningún
# otro municipio del país comparte ese nombre normalizado), así que tampoco
# hace falta ninguna pista geográfica adicional para saber su provincia.
_provincias_por_municipio_nacional = ine_historico.groupby("municipio_norm")["provincia_ine"].unique()
_municipio_unico_a_provincia = {
    municipio: provincias[0]
    for municipio, provincias in _provincias_por_municipio_nacional.items()
    if len(provincias) == 1
}

_pendientes_muni = todas.index[todas["total_poblacion"].isna()]
_resueltos_municipio_unico = 0
for idx in _pendientes_muni:
    municipio_norm = todas.at[idx, "municipio_norm"]
    prov_ine = _municipio_unico_a_provincia.get(municipio_norm)
    if prov_ine is None:
        continue
    anyo_fila = todas.at[idx, "anyo_cruce"]
    fila = _lookup.get(municipio_norm, {}).get(anyo_fila, {}).get(prov_ine)
    if fila is None:
        continue
    todas.loc[idx, ["provincia_ine", "total_poblacion", "poblacion_h", "poblacion_d"]] = [
        prov_ine, fila["total_poblacion"], fila["poblacion_h"], fila["poblacion_d"]
    ]
    superficie_val = _superficie_lookup.get(municipio_norm, {}).get(prov_ine)
    if superficie_val is not None:
        todas.loc[idx, "superficie_km2"] = superficie_val
        todas.loc[idx, "densidad_poblacion"] = fila["total_poblacion"] / superficie_val
    _resueltos_municipio_unico += 1

# Penúltimo fallback: alias de municipio verificados a mano (EDA,
# investigación sobre las filas con municipio conocido pero sin ninguna
# provincia resuelta). No es una heurística de similitud de texto tipo
# Levenshtein -- con nombres de municipio tan cortos eso da demasiados
# falsos positivos, y muchos de los valores que quedan sin resolver ni
# siquiera son un municipio real (texto de scraping, eventos "virtuales",
# nombres de club...) -- sino una lista corta de nombres truncados,
# transcritos con una errata o escritos en otra grafía, comprobados uno a
# uno contra el nomenclátor antes de añadirlos aquí. Los dos casos
# ambiguos a nivel nacional (más de un municipio con ese nombre exacto)
# se han resuelto con el contexto de la propia carrera: "Sobrado dos
# Monxes" es siempre el monasterio homónimo en A Coruña, nunca la otra
# "Sobrado" que también existe en el nomenclátor; "Aldehermosa de
# Montizón" se corre dentro del "Condado de Jaén", municipio Montizón.
MUNICIPIO_ALIAS = {
    "AS PONTES": ("AS PONTES DE GARCIA RODRIGUEZ", 15),
    "ALTORRICO": ("ALTORRICON", 22),
    "VIMINAZO": ("VIMIANZO", 15),
    "CALDES DE MALQAVELLA": ("CALDES DE MALAVELLA", 17),
    "RICON DE LA VICTORIA": ("RINCON DE LA VICTORIA", 29),
    "BURGUILLO DEL CERRO": ("BURGUILLOS DEL CERRO", 6),
    "SIDANOM": ("SIDAMON", 25),
    "ALMATRE": ("ALMATRET", 25),
    "VALLS DE CARDOS": ("VALL DE CARDOS", 25),
    "POBRA DO CARAMINAL": ("A POBRA DO CARAMINAL", 15),
    "VILAMARTIN": ("VILAMARTIN DE VALDEORRAS", 32),
    "SOBRADO DOS MONXES": ("SOBRADO", 15),
    "ALDEHERMOSA DE MONTIZON": ("MONTIZON", 23),
    "RIVEIRA": ("RIBEIRA", 15),
    "SANTA UXIA DE RIVEIRA": ("RIBEIRA", 15),
    "SANTA UXIA DE RIBEIRA": ("RIBEIRA", 15),
    "CANGAS DO MORRAZO": ("CANGAS", 36),
    "FERROL (ESTEIRO)": ("FERROL", 15),
    # FIX (21/9): 15 nucleos/parroquias mas de municipio sin provincia asignada, verificados
    # uno a uno (Wikipedia/turismo.gal/prensa local vía busqueda web, no de
    # memoria) porque no aparecen como municipio propio en el nomenclátor.
    "RAZO": ("CARBALLO", 15),  # playa/parroquia de Carballo (A Coruña)
    "TROBAJO DEL CAMINO": ("SAN ANDRES DEL RABANEDO", 24),  # barrio de San Andrés del Rabanedo (León)
    "MAGALOFES": ("FENE", 15),  # parroquia de Fene (A Coruña)
    "O MILLADOIRO": ("AMES", 15),  # núcleo/parroquia de Ames (A Coruña)
    "ARCA - PEDROUZO": ("O PINO", 15),  # Arca y O Pedrouzo, parroquias de O Pino (A Coruña), etapa del Camino de Santiago
    "BENQUERENCIA - BARREIROS": ("BARREIROS", 27),  # Benquerencia es un lugar dentro del municipio de Barreiros (Lugo)
    "ZUMAYA": ("ZUMAIA", 20),  # grafía castellana antigua de Zumaia (Gipuzkoa)
    "ANGROIS": ("SANTIAGO DE COMPOSTELA", 15),  # parroquia de Santiago de Compostela (A Coruña)
    "ANGROIS - SANTIAGO": ("SANTIAGO DE COMPOSTELA", 15),
    "A POBRA DE SAN XIAO": ("LANCARA", 27),  # parroquia de Láncara (Lugo)
    "ACEA DE AMA": ("CULLEREDO", 15),  # núcleo de Culleredo (A Coruña)
    "PRADOLLANO": ("MONACHIL", 18),  # estación de esquí de Sierra Nevada, en Monachil (Granada)
    "MONTEFURADO": ("QUIROGA", 27),  # túnel romano de Montefurado, en Quiroga (Lugo)
    "MARENY DEBARRAQUETES": ("SUECA", 46),  # pedanía de Sueca (Valencia); ya documentado en la nota de la celda del cruce ("Mareny de Barraquetes")
    "VILASECA DE SOLCINA": ("VILA-SECA", 43),  # núcleo de Vila-seca (Tarragona)
}

# Último fallback: municipio compuesto por varios nombres a la vez (p.ej.
# "Cambre - Culleredo", "Narón, Neda, Fene e Ferrol", o "Vigo - R.C.
# Celta" con el nombre de un club colado en medio de un municipio real).
# Partimos el texto por los separadores habituales y probamos cada trozo
# contra `_provincias_por_municipio_nacional` (ya calculado arriba); los
# trozos que no sean un municipio real (nombre de club, de asociación...)
# no hacen match y se ignoran sin más. Si los trozos que sí matchean
# coinciden todos en una sola provincia -- aunque alguno sea ambiguo por
# sí solo, como "Sada", que existe tanto en A Coruña como en Navarra, con
# "Noia" al lado ya solo cabe la primera -- asignamos esa provincia. Solo
# la provincia, no la población: no sabemos a cuál de los municipios
# nombrados atribuírsela.
def _provincias_de_municipio_compuesto(municipio_norm):
    trozos = [t for t in re.split(r"\s*-\s*|\s*,\s*|\s+E\s+|\s+Y\s+", municipio_norm) if t]
    provincias_por_trozo = [
        set(_provincias_por_municipio_nacional[t])
        for t in trozos
        if t in _provincias_por_municipio_nacional.index
    ]
    if not provincias_por_trozo:
        return None
    interseccion = set.intersection(*provincias_por_trozo)
    return interseccion.pop() if len(interseccion) == 1 else None


_pendientes_final = todas.index[todas["total_poblacion"].isna() & todas["municipio_norm"].notna()]
_resueltos_alias = 0
_resueltos_compuesto = 0
for idx in _pendientes_final:
    municipio_norm = todas.at[idx, "municipio_norm"]

    alias = MUNICIPIO_ALIAS.get(municipio_norm)
    if alias is not None:
        municipio_real, prov_ine = alias
        anyo_fila = todas.at[idx, "anyo_cruce"]
        fila = _lookup.get(municipio_real, {}).get(anyo_fila, {}).get(prov_ine)
        if fila is not None:
            todas.loc[idx, ["provincia_ine", "total_poblacion", "poblacion_h", "poblacion_d"]] = [
                prov_ine, fila["total_poblacion"], fila["poblacion_h"], fila["poblacion_d"]
            ]
            superficie_val = _superficie_lookup.get(municipio_real, {}).get(prov_ine)
            if superficie_val is not None:
                todas.loc[idx, "superficie_km2"] = superficie_val
                todas.loc[idx, "densidad_poblacion"] = fila["total_poblacion"] / superficie_val
            _resueltos_alias += 1
            continue

    prov_ine = _provincias_de_municipio_compuesto(municipio_norm)
    if prov_ine is not None:
        todas.loc[idx, "provincia_ine"] = prov_ine
        _resueltos_compuesto += 1

# FIX (21/9): para las filas que ni siquiera el alias de
# municipio resuelve, usamos la propia 'comarca' de la fuente -- una comarca
# tradicional (gallega, catalana, aragonesa...) esta contenida siempre en una
# unica provincia, así que aunque no sepamos el municipio exacto (a menudo
# porque es una parroquia/núcleo que no existe como tal en el nomenclátor del
# INE) sí sabemos la provincia. No aporta poblacion (no hay un municipio
# concreto que cruzar), solo provincia_ine -- que es lo que luego rellena
# 'provincia' (texto) en el paso de mas abajo. Las 30 comarcas de esta lista
# son las que aparecen en los datos sin resolver a fecha de este fix -- se
# puede ampliar si aparecen mas en el futuro.
COMARCA_A_PROVINCIA = {
    "A Barbanza": 15, "A Barcala": 15, "A Coruña": 15, "A Mariña Central": 27,
    "Alto Gállego": 22, "Arzúa": 15, "Bages": 8, "Bajo Aragón": 44,
    "Bajo Aragón-Caspe / Baix Aragó-Casp": 50, "Bergantiños": 15, "Berguedà": 8,
    "Betanzos": 15, "Campo de Gibraltar": 11, "Costa del Sol Occidental": 29,
    "Ferrol": 15, "Gúdar-Javalambre": 44, "Hoya de Huesca / Plana de Uesca": 22,
    "Morrazo": 36, "Noia": 15, "O Baixo Miño": 36, "O Condado": 36,
    "O Salnés": 36, "O Sar": 15, "O Xallas": 15, "Ordes": 15, "Pontevedra": 36,
    "Santiago": 15, "Sierra de Cádiz": 11, "Vallès Occidental": 8,
    "Vallès Oriental": 8, "Vigo": 36,
}
_resueltos_comarca = 0
_pendientes_comarca = todas.index[todas["provincia_ine"].isna() & todas["comarca"].notna()]
for idx in _pendientes_comarca:
    prov_ine = COMARCA_A_PROVINCIA.get(todas.at[idx, "comarca"])
    if prov_ine is not None:
        todas.loc[idx, "provincia_ine"] = prov_ine
        _resueltos_comarca += 1

# Un par de casos sin comarca informada tampoco, donde 'municipio' es en
# realidad una entidad que abarca varios municipios reales (un consorcio) o
# ya es el nombre de una comarca escrito directamente -- verificados por
# busqueda web, no de memoria: los municipios de "Consorcio das Mariñas"
# (residuos, área de A Coruña) están todos en A Coruña; "La Litera/La
# Llitera" es una comarca íntegramente en Huesca.
PROVINCIA_SOLO_POR_TEXTO = {
    "CONSORCIO DAS MARINAS": 15,
    "LA LITERA/LA LLITERA": 22,
}
_resueltos_texto = 0
_pendientes_texto = todas.index[todas["provincia_ine"].isna() & todas["municipio_norm"].notna()]
for idx in _pendientes_texto:
    prov_ine = PROVINCIA_SOLO_POR_TEXTO.get(todas.at[idx, "municipio_norm"])
    if prov_ine is not None:
        todas.loc[idx, "provincia_ine"] = prov_ine
        _resueltos_texto += 1

print(f"Filas resueltas por comarca de la fuente (sin municipio exacto): {_resueltos_comarca}")
print(f"Filas resueltas por entidad multi-municipio verificada a mano: {_resueltos_texto}")

cobertura = 100 * todas["total_poblacion"].notna().mean()
print(f"Filas resueltas gracias al fallback de comunidad autónoma: {_resueltos_ccaa}")
print(f"Filas resueltas por nombre de municipio único en España: {_resueltos_municipio_unico}")
print(f"Filas resueltas por alias de municipio verificado a mano: {_resueltos_alias}")
print(f"Filas resueltas (solo provincia, sin población) por municipio compuesto: {_resueltos_compuesto}")
print(f"Filas con población cruzada: {todas['total_poblacion'].notna().sum()} de {len(todas)} ({cobertura:.1f}%)")
print(f"Filas con superficie/densidad cruzada: {todas['superficie_km2'].notna().sum()} de {len(todas)}")
print(f"Filas con provincia_ine resuelta (con o sin población): {todas['provincia_ine'].notna().sum()} de {len(todas)} "
      f"({100*todas['provincia_ine'].notna().mean():.1f}%)")

# FIX (21/9): 'provincia' (el texto) solo traia lo que ya daba la fuente --
# aunque arriba ya resolvemos provincia_ine (el codigo) con varios fallbacks
# (comunidad uniprovincial, municipio unico en Espana, alias a mano...), esos
# fallbacks nunca escribian de vuelta en 'provincia'. Ahora que provincia_ine
# esta resuelto todo lo que se puede resolver, completamos 'provincia' (texto)
# para cualquier fila que todavia la tenga vacia pero SI tenga provincia_ine
# -- asi 'provincia' refleja siempre lo mejor que sabemos, no solo lo que la
# fuente escribio literalmente.
_CODIGO_A_PROVINCIA = {}
for _key_norm, _nombre in PROVINCIA_CANONICAS.items():
    _codigo = PROVINCIA_A_CODIGO.get(normalizar(_nombre))
    if _codigo is not None:
        _CODIGO_A_PROVINCIA.setdefault(_codigo, _nombre)

_provincia_texto_vacia = todas["provincia"].isna() & todas["provincia_ine"].notna()
todas.loc[_provincia_texto_vacia, "provincia"] = todas.loc[_provincia_texto_vacia, "provincia_ine"].map(_CODIGO_A_PROVINCIA)
print(f"'provincia' completada a partir de provincia_ine ya resuelto: {_provincia_texto_vacia.sum()} filas")

_con_municipio = todas["municipio"].notna()
_aun_sin_provincia = _con_municipio & todas["provincia"].isna()
print(f"Filas con municipio conocido y SIN provincia (tras el fix): {_aun_sin_provincia.sum()} de {_con_municipio.sum()}")

# FIX (21/9): 'municipio' se quedaba con la
# grafia mas frecuente en los propios datos (celda de unificacion de
# grafias, mas arriba) -- que no siempre es la mas bonita, si la fuente
# que mas filas aporta para un municipio lo escribe en mayusculas y sin
# acentos (p.ej. CADIZ, CEUTA, ALCALA LA REAL), esa era la grafia que
# ganaba. Ahora que provincia_ine esta resuelto del todo, sustituimos
# 'municipio' por el nombre oficial del nomenclator del IGN/CNIG
# (municipio_oficial, anadido en Limpieza_INE.ipynb a partir de
# NOMBRE_ACTUAL) alli donde el municipio ORIGINAL (sin pasar por alias ni
# por el fallback de comarca/compuesto) cruza directamente con el INE --
# como el cruce se hace por (municipio_norm, provincia_ine) tal cual
# quedaron desde el principio de esta celda, una fila resuelta por alias
# (p.ej. Trobajo del Camino -> San Andres del Rabanedo) no encuentra la
# clave TROBAJO DEL CAMINO en el nomenclator y no se toca -- no queremos
# renombrar con un nombre que la fuente nunca escribio.
todas = todas.merge(
    ine_superficie[["municipio_norm", "provincia_ine", "municipio_oficial"]]
    .dropna(subset=["municipio_oficial"])
    .drop_duplicates(subset=["municipio_norm", "provincia_ine"]),
    on=["municipio_norm", "provincia_ine"],
    how="left",
)
_municipio_actualizado = todas["municipio_oficial"].notna() & (todas["municipio"] != todas["municipio_oficial"])
print(f"'municipio' sustituido por la grafia oficial del nomenclator (IGN/CNIG): {_municipio_actualizado.sum()} filas")
todas.loc[todas["municipio_oficial"].notna(), "municipio"] = todas.loc[todas["municipio_oficial"].notna(), "municipio_oficial"]
todas = todas.drop(columns=["municipio_oficial"])


Filas resueltas por comunidad uniprovincial (antes del cruce): 0
Filas cuyo año de carrera queda fuera de 2000-2025 (se cruza con el año disponible más cercano): 2561
Filas resueltas por comarca de la fuente (sin municipio exacto): 584
Filas resueltas por entidad multi-municipio verificada a mano: 22
Filas resueltas gracias al fallback de comunidad autónoma: 7133
Filas resueltas por nombre de municipio único en España: 250
Filas resueltas por alias de municipio verificado a mano: 271
Filas resueltas (solo provincia, sin población) por municipio compuesto: 1187
Filas con población cruzada: 42619 de 49953 (85.3%)
Filas con superficie/densidad cruzada: 42615 de 49953
Filas con provincia_ine resuelta (con o sin población): 46470 de 49953 (93.0%)
'provincia' completada a partir de provincia_ine ya resuelto: 8462 filas
Filas con municipio conocido y SIN provincia (tras el fix): 71 de 46514
'municipio' sustituido por la grafia oficial del nomenclator (IGN/CNIG): 6037 filas


**Nota (3a mejora, 21/9, ampliada):** Clàudia pidió ir más allá de resolver solo el bug de las 9 CCAA uniprovinciales: "si té municipi hauria de poder assignar-li provincia SEMPRE". Además de completar `provincia` (texto) desde `provincia_ine` en cuanto este último está resuelto (recupera 8.461 filas en total, no solo las que ya resolvía el cruce con el INE antes de este fix), se añadieron dos mecanismos más para resolver `provincia_ine` en los casos donde ni el municipio exacto ni sus alias bastaban: (1) 15 alias de municipio nuevos, verificados uno a uno por búsqueda web (no de memoria) contra Wikipedia/turismo.gal/prensa local -- parroquias y núcleos gallegos sobre todo (Razo, Angrois, Acea de Ama...), más Pradollano (Granada), Trobajo del Camino (León) y Vilaseca de Solcina (Tarragona); (2) un fallback por `comarca` (30 comarcas tradicionales, todas contenidas en una única provincia) para filas cuyo `municipio` es una parroquia o núcleo que no existe como tal en el nomenclátor del INE pero cuya comarca sí basta para saber la provincia -- y dos entidades multi-municipio verificadas a mano ("Consorcio das Mariñas", "La Litera/La Llitera") resueltas igual, por texto. Ninguno de estos dos mecanismos aporta población (no hay un municipio concreto que cruzar con el INE), solo `provincia_ine` -- que es justo lo que necesita `provincia` (texto). Quedan 71 filas (0,15% de las que tienen `municipio`) sin resolver, todas verificadas como genuinamente irresolubles: o bien no son un lugar real ("Varios On Line", "Cualquier lugar del mundo", "Virtual", "España", "Prueba", "Campionato", "Igasoft"), o bien son ambiguas incluso a nivel de provincia ("Cobas" existe en A Coruña Y en Ourense; "A Pobra" coincide con varios municipios "A Pobra de X" en provincias distintas; "A Silva" es un topónimo demasiado genérico; "Ribagorza" es una comarca histórica repartida entre Huesca y Lleida).

In [13]:
print("Cobertura del cruce con el INE por fuente (% de filas con población encontrada):")
print(todas.groupby("fuente")["total_poblacion"].apply(lambda s: round(100 * s.notna().mean(), 1)))
print()

print("Valores de 'provincia' sin equivalente en el INE (top 10, casi todos son comunidades autónomas):")
print(todas.loc[todas["provincia"].notna() & todas["provincia_ine"].isna(), "provincia"].value_counts().head(10))

Cobertura del cruce con el INE por fuente (% de filas con población encontrada):
fuente
buscametas          72.5
carreirasgalegas    89.3
ccnorte             64.6
championchip        82.4
cronofinisher       92.0
cronorunner         87.4
cruzandolameta      91.2
cursescat           89.9
iter5               79.8
mychip              75.5
sportmaniacs        95.4
timingsys           67.3
youevent            54.4
Name: total_poblacion, dtype: float64

Valores de 'provincia' sin equivalente en el INE (top 10, casi todos son comunidades autónomas):
Series([], Name: count, dtype: int64)


Antes de mejorar el cruce, la cobertura era baja sobre todo en `ccnorte` (27.7%), `mychip` (52.5%) y `championchip` (63.4%), casi siempre por el mismo motivo: la columna `provincia` guarda en realidad el nombre de la comunidad autónoma (Galicia, Catalunya, Comunidad Valenciana...), y una comunidad con varias provincias no se puede asignar directamente a un código INE sin más información. Añadimos cuatro mejoras al cruce (celda de arriba): (1) cuando `provincia` es una comunidad multiprovincial, buscamos el municipio dentro de las provincias que la componen y solo lo damos por válido si el nombre es único en toda la comunidad, para no arriesgarnos a cruzar con la provincia equivocada; (2) tratamos como equivalentes las dos mitades de los nombres oficiales dobles del nomenclátor, tanto en el municipio (p.ej. "Alboraia/Alboraya") como en la propia `provincia` (p.ej. "València / Valencia", "Alacant / Alicante"); (3) el nomenclátor pospone el artículo entre paréntesis ("CORUÑA (A)", "PINO (O)", "ALCÚDIA (L')") — lo pasamos delante, que es como se escribe en las fuentes ("A Coruña", "O Pino", "l'Alcúdia"), cuidando de no meter un espacio de más cuando el artículo lleva apóstrofo elidido; y (4) esa misma conversión del artículo pospuesto, que al principio solo se aplicaba al nombre de las carreras, ahora se aplica también a `municipio_norm` del lado del INE (`ine_historico`/`ine_superficie`) — sin esto, ~15.500 filas del nomenclátor se quedaban con el nombre en su forma original ("POBLA DE FARNALS (LA)") y nunca cruzaban contra la forma con el artículo delante que traen las fuentes, sobre todo en la Comunidad Valenciana, donde este formato es muy común. Con esto la cobertura global sube del ~73% al 85.6%; en concreto `ccnorte` pasa a 64.4%, `mychip` a 75.0%, `championchip` a 82.4% y `sportmaniacs` (el más beneficiado por el punto 4, con la mayoría de nombres tipo "(LA)"/"(EL)" en su catálogo) a 95.4%. El resto de la brecha es sobre todo el mismo municipio escrito con nombres distintos: exónimos históricos en castellano frente al nombre oficial actual ("Alcira" vs "Alzira", "Játiva" vs "Xàtiva" en mychip), nombres truncados y ambiguos entre varios municipios de una misma comunidad (p.ej. "Sant Llorenç" a secas, que coincide con 4 municipios distintos de Catalunya), municipios sin el artículo que sí lleva su nombre oficial (p.ej. "Pobla de Vallbona" por "La Pobla de Vallbona"), entidades locales menores que no son un municipio propio en el nomenclátor (p.ej. "Mareny de Barraquetes", pedanía de Sueca), y al menos un caso de `provincia` mal etiquetada en origen (carreras de Barbastro, Huesca, marcadas como `Catalunya` en `iter5`). Cerrar esto ya requeriría una tabla de sinónimos hecha a mano o revisar la fuente caso a caso — queda como posible ampliación futura.

`cronorunner` llega aquí con `municipio`/`comarca`/`provincia` ya geocodificados de verdad en `Limpieza_cronorunner.ipynb` (con Nominatim, sobre el municipio real cuando lo hay y si no sobre un candidato extraído del nombre del evento), así que cruza con el INE igual de bien que el resto: 85.9% de cobertura, por encima incluso de la media global. El hueco que queda es la misma causa que ya limita el propio `municipio` de origen: en 1.605 de sus 3.145 filas (51%) el scraper marcó `municipio` como `"Sin Localidad definida"` porque la ficha del evento no traía icono de ubicación, y ahí la geocodificación por candidato de nombre no siempre encuentra nada real que cruzar (149 de las 754 consultas únicas a Nominatim, entre municipios reales y candidatos, se quedaron sin resultado). A partir de aquí, cualquier cálculo con `total_poblacion` trabaja solo sobre ese subconjunto cruzado, no sobre la tabla entera.

**Nota (regresión detectada y corregida):** `normalitzar_ubicacions()` vacía `provincia` en cuanto detecta que el valor era en realidad una comunidad autónoma y lo traslada a `comunidad_autonoma` — pero el fallback multiprovincial de arriba solo miraba `provincia_norm`, nunca `comunidad_autonoma_norm`, así que cualquier fuente cuya `provincia` trajera directamente el nombre de la comunidad (el caso de `cronorunner`, con "Comunitat Valenciana" en el 85% de sus filas) se quedaba prácticamente sin cruzar: `cronorunner` había caído a 1.5% de cobertura y el global a 62.2%, muy por debajo de lo que documenta este mismo texto. Se han corregido tres cosas: (a) el fallback multiprovincial ahora también mira `comunidad_autonoma_norm`; (b) se añadió un mapeo directo, sin ambigüedad posible, para las comunidades uniprovinciales (Madrid, Asturias, Cantabria, Murcia, Navarra, La Rioja, Illes Balears, Ceuta, Melilla), que antes se quedaban sin cruzar por el mismo motivo aunque no necesiten ninguna comprobación de unicidad; y (c) un último fallback asigna la provincia cuando el nombre del municipio es único en todo el nomenclátor de España, para las pocas filas que no traen ninguna pista de provincia ni comunidad. Con esto la cobertura global sube a 84.9% (`cronorunner` 87.4%, `sportmaniacs` 95.4%, `championchip` 82.4% — de nuevo en línea con los números de este texto) y ya no queda ningún valor de `provincia` sin equivalente en el INE que no sea ya una comunidad autónoma genuina.

### `es_carrera_popular`: identificar carreras populares por el nombre

Las carreras populares ("carrera/cursa/carreira popular") son un tipo de evento que se reconoce por su propio nombre, en las 3 lenguas de las fuentes: pensadas para participación masiva y abierta a todo el mundo, no para competición federada. Además de la palabra "popular" en sí, ampliamos la detección a otros patrones de nombre igual de reconocibles y validados contra los datos antes de incluirlos:

- **"San Silvestre"/"Sant Silvestre"**: el formato más estandarizado que hay — carrera de Nochevieja abierta a todo el mundo, prácticamente siempre popular por definición. Se amplía para cubrir también la grafía catalana (`Sant Silvestre de Berga`, `Sant Silvestre del Masnou`...), que antes se quedaba fuera.
- **"solidari(a/o)"**/**"benefic(a/o)"**: carreras solidarias/benéficas — abiertas y participativas, no competitivas (`Cursa Benèfica Malalties Minoritàries`, `Travesía a Nado Benéfica`...).
- **"festa/festes/fiesta"** (con límite de palabra, para no capturar "manifestación" y similares): carreras de fiesta mayor/patronal.
- **"nocturn[oa]"/"night"**: carreras nocturnas, en las 2 lenguas y también en inglés (`Night Race`, `Urban Trail Night`...) — el patrón anterior solo cogía "nocturna" en femenino y se dejaba fuera "nocturno" y los nombres en inglés.
- **"mujer"/"dona"**: carreras de la mujer/cursa de la dona — eventos participativos abiertos, casi siempre en categoría Absoluta/General.
- **"milla urbana"**: formato clásico de evento popular de ciudad, sobre todo en categorías infantiles.
- **"legua"**: unidad de distancia tradicional que da nombre a carreras populares de pueblo (`Legua Irmandiña`, `Legua Ciudad de Valladolid`...).

Además, el patrón se aplica ahora sobre el nombre **sin acentos** (reutilizando `_normalizar_nombre`, ya definida más arriba para el cruce de duplicados) en vez de sobre el nombre tal cual: el patrón `solidari` no estaba pillando la grafía catalana `SOLIDÀRIA` por la tilde, y esa sola diferencia dejaba fuera 242 filas que sí eran carreras solidarias reales.

Se descartaron `san`/`sant` a secas (demasiado ruidoso: mezcla carreras patronales populares con pruebas serias tituladas "San/Sant `<algo>`", el mismo problema por el que ya se había descartado "memorial") y `trofeo`/`nacional`/`circuito`/`premio` (van justo en la dirección contraria: son señal de evento competitivo/federado, no de carrera popular).

Con todos los patrones (incluido `benefic`, añadido después de `solidari` -- cubre 69 filas más, como "Cursa Benèfica Malalties Minoritàries" o "Travesía a Nado Benéfica", que `solidari` no cogía), el índice detecta 12.393 filas (25.3% del total), y sigue encajando con el patrón esperado: sobre todo `road running`, y `publico` Absoluta/General o Infantil.

In [14]:
PATRON_POPULAR = (
    r"\bpopular|\bsant? silvestre\b|\bsolidari|\bbenefic|\bfiesta\b|\bfestes?\b"
    r"|\bnocturn[oa]\b|\bnight\b|\blegua\b|\bmujer\b|\bdona\b|\bmilla urbana\b"
)
_nombre_sin_acentos = todas["nombre_carrera"].apply(_normalizar_nombre)
todas["es_carrera_popular"] = _nombre_sin_acentos.str.contains(PATRON_POPULAR, case=False, na=False, regex=True)

print(f"Carreras populares detectadas: {todas['es_carrera_popular'].sum()} de {len(todas)} ({100 * todas['es_carrera_popular'].mean():.1f}%)")
print()
print("Por fuente:")
print(todas.loc[todas["es_carrera_popular"], "fuente"].value_counts())
print()
print("Por tipo_modalidad:")
print(todas.loc[todas["es_carrera_popular"], "tipo_modalidad"].value_counts())
print()
print("Por publico:")
print(todas.loc[todas["es_carrera_popular"], "publico"].value_counts())

Carreras populares detectadas: 12689 de 49953 (25.4%)

Por fuente:
fuente
sportmaniacs        4381
carreirasgalegas    2650
ccnorte             2081
mychip               768
cronorunner          630
cruzandolameta       533
championchip         484
youevent             456
cronofinisher        431
iter5                103
timingsys            100
cursescat             52
buscametas            20
Name: count, dtype: int64

Por tipo_modalidad:
tipo_modalidad
Road running       11198
Trail running        648
Marcha               388
Otros                246
Ciclismo y btt        73
Multidisciplina       69
Natación              67
Name: count, dtype: int64

Por publico:
publico
Absoluta/General     7552
Infantil             4759
Otros                 175
Mayores/Veteranos     148
Equipos                39
Elite                  16
Name: count, dtype: int64


## Guardado final

Guardamos la tabla ya unida, con las variables derivadas, el cruce con población/superficie/densidad por año y el índice de carrera popular, lista para `Analisis_union.ipynb` y para cualquier futuro notebook de modelado. `DF_TODAS.csv` (guardado más arriba, sin estas columnas) se mantiene igual por compatibilidad con los notebooks que ya lo usan directamente.

In [15]:
SALIDA_UNION = Path("../../data/processed/union/DF_TODAS_UNION.csv")
todas.to_csv(SALIDA_UNION, index=False, encoding="utf-8-sig")
print(f"Guardado en {SALIDA_UNION}")
print("Columnas:", list(todas.columns))

Guardado en ../../data/processed/union/DF_TODAS_UNION.csv
Columnas: ['fuente', 'nombre_carrera', 'fecha', 'dia_semana', 'distancia', 'tipo_modalidad', 'publico', 'finisher_d', 'finisher_h', 'municipio', 'comarca', 'provincia', 'finisher_desconocido', 'comunidad_autonoma', 'anyo', 'mes', 'estacion', 'total_finishers', 'es_finde', 'municipio_norm', 'provincia_norm', 'comunidad_autonoma_norm', 'provincia_ine', 'anyo_cruce', 'total_poblacion', 'poblacion_h', 'poblacion_d', 'superficie_km2', 'densidad_poblacion', 'es_carrera_popular']
